### ISSUE TITLE
Tips on improving the model finetuning (not an issue)

### ISSUE THREAD

**Issue by fattynoparents** [knaw-huc/loghi#38]
I've been finetuning my transcription model for some time now. 

The workflow is the following: I take the data -> split it manually into training and validation sets -> train my model -> run it on unseen material -> prepare new data -> add it to the data the model has already been trained on -> repeat the whole process.

Right now I have the training material of about 1300 images with 1 to 50 textlines on each, with 1 to 5 various handstyles on each page, with mixed handwritten and typed text on many images.

I have noticed that at some point when I check the model performance on unseen material it started becoming better at some textlines while becoming worse at others, and then with the next training round it becomes better where it was not so good before, but becomes worse where it used to be good. And so the overal performance is not improving.

Can this happen due to the way I split the data? So that some handstyles become overrepresented and the model concentrates on them while forgetting what it learned in the previous run? Are there any tips how to avoid this?

Here's also the config I use when finetuning:
```
{
    "args": {
        "augmentation": {
            "aug_binarize_otsu": false,
            "aug_binarize_sauvola": false,
            "aug_blur": false,
            "aug_distort_jpeg": false,
            "aug_elastic_transform": true,
            "aug_invert": false,
            "aug_multiply": 1,
            "aug_random_crop": true,
            "aug_random_shear": false,
            "aug_random_width": true,
            "visualize_augments": false
        },
        "decoding": {
            "beam_width": 10,
            "greedy": false
        },
        "general": {
            "batch_size": 32,
            "gpu": "0",
            "output": "output",
            "seed": 42
        },
        "learning_rate": {
            "decay_per_epoch": false,
            "decay_rate": 0.95,
            "decay_steps": -1,
            "learning_rate": 0.0003,
            "linear_decay": true,
            "optimizer": "rmsprop",
            "warmup_ratio": 0.0
        },
        "model": {
            "freeze_conv_layers": true,
            "freeze_dense_layers": false,
            "freeze_recurrent_layers": false,
            "model": "/path/to/model/",
            "model_name": "My-Finetuned-Loghi-Model",
            "replace_final_layer": true,
            "replace_recurrent_layer": null,
            "use_float32": false
        },
        "training": {
            "do_validate": true,
            "early_stopping_patience": 20,
            "epochs": 100,
            "max_queue_size": 256,
            "output_checkpoints": true,
            "train_list": "/path/to/train.txt",
            "training_verbosity_mode": "auto",
            "validation_list": "/path/to/val.txt"
        }
    }
}
```

**Comment by rvankoert** [comment]
It can happen that some lines get a bit worse and others better, but in general it should improve overall. Especially after several epochs. It could be an issue on how the data is split, but i doubt it should have that much impact. We know from projects that a lot of knowledge from the foundation models is retained even after finetuning for many epochs on different data.

I prefer to use random splitting, either by page if you have a lot of data. Or by line (the default) if you have little data.

First try:
- lower your learning rate to 0.0001

Another trick to try:
If you have a lot of unlabeled data:
- train a model
- inference as much unseen data as possible (or just a couple of thousand to test this method)
- edit the create-train-data.sh 
-- replace `-no_page_update \` with `-no_page_update -minimum_confidence 0.8 \`
- run the create-train against the just inferenced data and output to some "pseudo-data"-folder
- train 1 or two epochs on the pseudo-data and your actual ground truth
- (optional) finetune against your ground truth (do this when you have very little pseudo data)
You should get a more robust model that also has lower CER. Try playing with the 0.8 value.

**Comment by rvankoert** [comment]
in addition: try
"freeze_conv_layers": false

Only use this
"replace_final_layer": true
if you are adding/removing characters from the possible outputs

**Comment by fattynoparents** [comment]
Wow, thanks a lot for the tips, I really appreciate it! I will try and let you know if it helps.

**Comment by fattynoparents** [comment]
Ok so I have now tried all the suggested variants but unfortunately didn't get any better results. I suppose my model just reached the upper border of learning, at least Loghi shows VAL_CER of about 0.03.
It is pretty good on most of the unseen material but it still is at times better at times worse on the 50 test images we use as a control.
I will see if increasing the amount of training material will maybe provide any better results.
